### PR TITLE
Add initial support for the Dual Motor TinyShield to the Propulsion Module

### PR DESCRIPTION
**Pull request by whymarrh**
This PR adds support for the [TinyCircuits Dual Motor TinyShield](https://tinycircuits.com/collections/all/products/dual-motor-tinyshield) by conditionally replacing usages of `PololuQik2s12v10` with the [TinyCircuits Motor Drvier Arduino library](https://github.com/TinyCircuits/TinyCircuits-TinyShield_Motor_Library)'s `MotorDriver`.

My thinking is that Dual Motor TinyShield is the special case here and that it should be the check performed (i.e. "if this is being built for a TinyDuino do *x*, otherwise the default is an Arduino" vs. "if this is being built for an Arduino do *x*, otherwise the default is a TinyDuino") and as such I've added `#ifdef`s for a `TINYDUINO` macro where needed.

One complication is that a lot of the functionality exposed by the Arduino library for the Pololu Qik isn't available with the TinyShield. I've proposed a new payload value for the Propulsion Module API's General Error Message ([`0x1F`][2]) to indicate that the given command isn't available—I'm very open to alternatives here, as I don't think this will stand the test of time (e.g. when we go to use these commands we'll need them to work equally across both boards).

  [2]:https://github.com/LakeMaps/microcontrollers/wiki/Propulsion-Microcontroller-API#command-0x1f--general-error-message

I've tested this with the hardware with the following script and it seems to work well:

```text
>>> import serial
>>> s = serial.Serial('/dev/cu.usbserial-DO00BW3D', baudrate=57600)
>>> s.write(bytearray([0xaa, 0x13, 0x00, 0x7f, 0x00, 0x00, 0x79, 0x9b]))
8
>>> hex(ord(s.read()))
'0xaa'
>>> hex(ord(s.read()))
'0x13'
>>> hex(ord(s.read()))
'0x0'
>>> hex(ord(s.read()))
'0x7f'
>>> hex(ord(s.read()))
'0x0'
>>> hex(ord(s.read()))
'0x0'
>>> hex(ord(s.read()))
'0x79'
>>> hex(ord(s.read()))
'0x9b'
>>> s.write(bytearray([0xaa, 0x13, 0x00, 0x7f, 0x00, 0x00, 0x79, 0x9b]))
8
```

### CODE REVIEW

**Comment by whymarrh** [comment]
> I've tested this with the hardware with the following script and it seems to work well:

Actually, that's not quite true. I have it working with 1/2 motors—the 2nd motor is not yet connected.

File: Arduino/Propulsion/src/PropulsionModule.cpp [hunk 1]
```diff
@@ -1,6 +1,11 @@
 #include <Arduino.h>
+#ifdef TINYDUINO
```
Comment (arandell93) [inline Arduino/Propulsion/src/PropulsionModule.cpp]: Is there are more explicit way to group these, brackets or similar? I found it hard to follow at first. Perhaps if nothing exists we can indent the conditional lines to make it more apparent?

File: Arduino/Propulsion/src/PropulsionModule.cpp [hunk 2]
```diff
@@ -54,7 +70,7 @@ int16_t byteToInt(int16_t index);
 void setup() {
   Serial.begin(SERIAL_BAUD);
   Serial.setTimeout(serialTimeout);
-  qik.init();
+  initMotor();
```
Comment (arandell93) [inline Arduino/Propulsion/src/PropulsionModule.cpp]: What was your reasoning for moving away from a call to the library directly?
Further, initMotor is a poor name as you aren't initializing a motor, you're initializing the driver board itself, which then 'arms' two motors. At the same time, I don't like something like initController, as that is ambiguous as to whether it means microcontroller or motor controller. Maybe something like initDriver?

File: Arduino/Propulsion/src/PropulsionModule.cpp [hunk 3]
```diff
@@ -36,7 +41,18 @@ uint16_t byteToShort(int16_t index);
 void errorReply(uint8_t errorByte);
 
 /*****Propulsion Module Specific Parameters**************************/
+#ifdef TINYDUINO
+MotorDriver motor(0);
+void initMotor() {
+  motor.begin(127);
```
Comment (arandell93) [inline Arduino/Propulsion/src/PropulsionModule.cpp]: Does the motor speed range from 0-255 with 127 being zero speed?

File: Arduino/Propulsion/src/PropulsionModule.cpp [hunk 4]
```diff
@@ -36,7 +41,18 @@ uint16_t byteToShort(int16_t index);
...
+  motor.setFailsafe(1000);
```
Comment (arandell93) [inline Arduino/Propulsion/src/PropulsionModule.cpp]: Is this a timeout? Would it be more prudent to enumerate it as such?

File: Arduino/Propulsion/src/PropulsionModule.cpp [hunk 5]
```diff
@@ -139,6 +155,10 @@ void ResetModule() {
 }
 
 void GetConfig() {
+#ifdef TINYDUINO
+  errorByte = 0x03;
```
Comment (arandell93) [inline Arduino/Propulsion/src/PropulsionModule.cpp:167]: I like this concept, and I think it would be a good idea to extend it to support the qik as well. That is, you can't get the config of a register that isn't a valid configuration register, so perhaps we can perform a check on the payload to see what register is being requested and compare it to see if it is valid, else return error 0x03.

**Review by arandell93 — COMMENTED** [review]
Few questions regarding extensibility of changes for qik, general comments of readability/ambiguity.

File: Arduino/Propulsion/src/PropulsionModule.cpp (hunk 1 above)
Comment (whymarrh) [reply Arduino/Propulsion/src/PropulsionModule.cpp]: I don't often see preprocessor directives indented, so I'm hesitant to do that. I could add extra newlines?

```cpp
#include <Arduino.h>

#ifdef TINYDUINO

#include <Wire.h>
#include <MotorDriver.h>

#else

#include <SoftwareSerial.h>
#include <PololuQik.h>

#endif

#include <Crc16.h>
```

Comment (whymarrh) [reply Arduino/Propulsion/src/PropulsionModule.cpp]: That said, preprocessor directives are generally weird and initially-hard-to-follow (and some may say that C++ as a language is generally so).

File: Arduino/Propulsion/src/PropulsionModule.cpp (hunk 3 above)
Comment (whymarrh) [reply Arduino/Propulsion/src/PropulsionModule.cpp]: I'll double-check this—127 here should indicate the maximum PWM value

Comment (whymarrh) [reply Arduino/Propulsion/src/PropulsionModule.cpp]: It should also be a constant, so I'll do that

File: Arduino/Propulsion/src/PropulsionModule.cpp (hunk 4 above)
Comment (whymarrh) [reply Arduino/Propulsion/src/PropulsionModule.cpp]: Yes? "The failsafe turns off motors if a command is not sent in a certain amount of time." It can be disabled apparently. Does the Qik 2s12v10 timeout?

File: Arduino/Propulsion/src/PropulsionModule.cpp (hunk 2 above)
Comment (whymarrh) [reply Arduino/Propulsion/src/PropulsionModule.cpp]: I'm open to either renaming (e.g. `initDriver` or `initController`) or removing this indirection. What makes more sense to you?

File: Arduino/Propulsion/src/PropulsionModule.cpp (hunk 5 above)
Comment (whymarrh) [reply Arduino/Propulsion/src/PropulsionModule.cpp:167]: I agree that it can be extended to handle invalid requests for Qik registers but we can save that for a different PR to keep things localized here.

**Comment by whymarrh** [comment]
I'm still working out what needs to be passed to `MotorDriver::begin`, but I can confirm that `MAX_PWM` set to 127 and using -127, 0, and 127 for speeds works for the basic forward and reverse:

```python
# Forwards M1
s.write(bytearray([0xaa, 0x13, 0x00, 0x7f, 0x00, 0x00, 0x79, 0x9b]))
# Forwards M2
s.write(bytearray([0xaa, 0x13, 0x00, 0x00, 0x00, 0x7f, 0x02, 0xda]))
# Reverse M1
s.write(bytearray([0xaa, 0x13, 0xFF, 0x81, 0x00, 0x00, 0xCA, 0x6B]))
# Reverse M2
s.write(bytearray([0xaa, 0x13, 0x00, 0x00, 0xff, 0x81, 0x0f, 0xf4]))
```

File: Arduino/Propulsion/src/PropulsionModule.cpp (hunk 3 above)
Comment (whymarrh) [reply Arduino/Propulsion/src/PropulsionModule.cpp]: From what I understand, this is indeed what value should represent max forward speed, so 127 is correct

File: Arduino/Propulsion/src/PropulsionModule.cpp (hunk 1 above)
Comment (arandell93) [reply Arduino/Propulsion/src/PropulsionModule.cpp]: New Lines are good enough.

File: Arduino/Propulsion/src/PropulsionModule.cpp (hunk 2 above)
Comment (arandell93) [reply Arduino/Propulsion/src/PropulsionModule.cpp]: Let's do initDriver

File: Arduino/Propulsion/src/PropulsionModule.cpp (hunk 5 above)
Comment (arandell93) [reply Arduino/Propulsion/src/PropulsionModule.cpp:167]: Sounds like a plan. We'll add it to the next revision we do for the Qiks once we get to the stage where we are using the extra features.

File: Arduino/Propulsion/src/PropulsionModule.cpp (hunk 2 above)
Comment (whymarrh) [reply Arduino/Propulsion/src/PropulsionModule.cpp]: Alrighty

**Comment by whymarrh** [comment]
This should be good for another round of review

**Comment by whymarrh** [comment]
I'll update the General Error Message response payload table to look like so:

Value|	Description|
|:---:|:---:|
0x01|CRC Error
0x02|Speed Command out of Range
0x03|Not Applicable

**Review by arandell93 — APPROVED** [review]
